File: migratory-birds/migratory-birds.js
// https://www.hackerrank.com/challenges/migratory-birds/problem

const pipe = (...fns) => initialData =>
  fns.reduce((updatedData, fn) => fn(updatedData), initialData);

const reduce = (fn, startingVal) => arr => arr.reduce(fn, startingVal);

const migratoryBirds = ids =>
  pipe(
    reduce((ret, id) => {
      ret[id] = id in ret ? ret[id] + 1 : 1;
      return ret;
    }, {}),
    idToTotalSightings => {
      const maxSightings = Math.max(...Object.values(idToTotalSightings));

      const [id] = Object.entries(idToTotalSightings).find(
        ([id, totalSightings]) => totalSightings === maxSightings
      );

      return id;
    }
  )(ids);

File: migratory-birds/migratory-birds.clj
(defn max-k [id-to-total]
  (let [max-total (apply max (vals id-to-total))]
    (->>  (filter
           (fn [id] (= max-total (id-to-total id)))
           (keys id-to-total))
          (apply min))))

(defn migratoryBirds [ids]
  (->>  (reduce
         (fn [ret id]
           (update ret id (fn [v] (if (ret id) (inc (ret id)) 1))))
         {}
         ids)
        max-k))